File: src/hello_ros/hello_ros/talker.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String          # 消息类型：std_msgs 包里的 String


class Talker(Node):

    def __init__(self):
        super().__init__('talker')
        self.count = 0
        self.pub = self.create_publisher(String, 'chatter', 10)
        self.timer = self.create_timer(1.0, self.tick)

    def tick(self):
        self.count += 1

        msg = String()
        msg.data = f'第 {self.count} 次心跳'
        self.pub.publish(msg)

        self.get_logger().info(f'发布: {msg.data}')


def main(args=None):
    rclpy.init(args=args)
    node = Talker()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/hello_ros_cpp/src/talker.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;


class Talker : public rclcpp::Node
{
public:
    Talker() : Node("talker_cpp")        // ≈ super().__init__('talker_cpp')
    {
        count_ = 0;

        // 类型写在 <> 里 —— C++ 编译期就得知道发布的是什么消息
        pub_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
        timer_ = this->create_wall_timer(1s, [this]() { tick(); });
    }

private:
    void tick()
    {
        count_ += 1;

        auto msg = std_msgs::msg::String();
        msg.data = "第" + std::to_string(count_)+"次心跳";              // ⚠️ 待补：换成真正的中文内容
        pub_->publish(msg);

        RCLCPP_INFO(this->get_logger(), "发布：%s", msg.data.c_str());   // ⚠️ 待补：同上
    }

    // 成员变量的类型 = 初始化它的那个函数的返回值类型
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    int count_;
};


int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);                       // ≈ rclpy.init()
    rclcpp::spin(std::make_shared<Talker>());       // ≈ rclpy.spin(node)
    rclcpp::shutdown();                             // ≈ rclpy.shutdown()
    return 0;
}
